### ISSUE TITLE
[Review/Improve] fail2ban service

### ISSUE THREAD

**Issue by casperklein** [docker-mailserver/docker-mailserver#1821]
# Review/Improve Request

## Context

At the moment, the fail2ban service seems not to operate optimal.

### Is your Feature Request related to a Problem?

At least:

https://github.com/docker-mailserver/docker-mailserver/pull/1661
https://github.com/docker-mailserver/docker-mailserver/issues/1810

### Describe the Solution you'd like

The whole fail2ban implementation should be reviewed and optimized.

### Are you going to implement it?

No (at least, not in short-term). This is more for tracking/discussing the topic.

### What are you going to contribute?

Lifetime 😄

# Things to discuss / clarify / To-do

- ~~Is `https://github.com/docker-mailserver/docker-mailserver/blob/master/target/fail2ban/jail.conf` still needed, what was the original intent not to use the default shipped configuration?~~ #1895 
- ~~What services are covered by fail2ban?~~ dovecot, postfix, postfix-sasl
- ~~Should an attacker be blocked completely? Currently, the IP address is only blocked for a specific service. IMHO if someone tries to brute force POP3, the IP address should also be blocked for IMAP etc, or even better completely.~~
- ~~Default ban behavior of fail2ban is to REJECT packets. Maybe DROP is better?~~
- ~~Check if `update-alternatives --set iptables /usr/sbin/iptables-legacy # switch iptables and ip6tables to legacy for Fail2Ban` from Dockerfile is (still) needed~~

There are probably other things to consider. Just let me know and I'll keep this list up-to-date.

### Block all ports instead of service port only
`sed -i '/^banaction = iptables-multiport$/ s|iptables-multiport|iptables-allports|' /etc/fail2ban/jail.conf`

### DROP connections instead of REJECT
`sed -i 's/^blocktype = REJECT.*/blocktype = DROP/' /etc/fail2ban/action.d/iptables-common.conf`

### Proposed configuration options

| Variable | Values | Default | Description |
| --- | --- | ---  | --- |
| FAIL2BAN_BLOCK_TYPE | REJECT, DROP | DROP | iptables block behaviour; `REJECT` sends *ICMP unreachable*; `DROP ` sends no reply |
| FAIL2BAN_BLOCK_ALL | true, false | true | If *true*, block an attacker on all ports, not only the attacked one. |
| FAIL2BAN_BLOCK_PERMANENT | true, false | false | Block an attacker forever |

**Comment by aendeavor** [comment]
> Default ban behavior of fail2ban is to REJECT packets. Maybe DROP is better?

Agree.

> Should an attacker be blocked completely? Currently, the IP address is only blocked for a specific service. IMHO if someone tries to brute force POP3, the IP address should also be blocked for IMAP etc, or even better completely.

Completely from my PoV.

> What services are covered by fail2ban?

No idea.

> Is https://github.com/docker-mailserver/docker-mailserver/blob/master/target/fail2ban/jail.conf still needed, what was the original intent not to use the default shipped configuration?

No idea. But after all these years, we could assume that we can completely revamp behavior :D

**Comment by aendeavor** [comment]
I can confirm that F2B is not working 100%. I was seeing many login attempts today on my server which should have been blocked - but they weren't.

**Comment by polarathene** [comment]
Pretty sure my linked PR provides some insights into the history and raised similar questions posed here. I'll be happy to look into it further when time permits, possibly next month. I'll diff the conf against the current upstream to see if was anything specific that stands out if `git blame` doesn't reveal anything useful.

I am not too knowledgeable about F2B myself, but any bans should not be permanent. Sometimes IP can be shared across multiple systems and IIRC spoofed, so you'd possibly want to be careful of Denial of Service (to users with same IP).

Also related to my PR, if your bans are related to IMAP it could be due to improper config as I point out:

> *I've not verified, but presumably `imap` in the current config for Dovecot, shouldn't match port 143 and thus not trigger Fail2Ban, only `imaps` on port 993.*

Which is misconfigured due to changing the base image, which as I point out different distros can have different service mappings and this was missed (since we carried a non-upstream config over).

**Comment by casperklein** [comment]
> I am not too knowledgeable about F2B myself, but any bans should not be permanent

The default ban time is afaik 10 minutes and wasn't subject to change?

**Comment by BlockI0tChain** [comment]
>  I am not too knowledgeable about F2B myself, but any bans should not be permanent


A permanent ban should be nevertheless possible, as an option at least.

**Comment by casperklein** [comment]
I've added possible configuration options to my first post, including `BLOCK_PERMANENT`.

**Comment by aendeavor** [comment]
> I've added possible configuration options to my first post, including `BLOCK_PERMANENT`.

SGTM 👍🏼
I really like this!

**Comment by aendeavor** [comment]
What's the status here?

**Comment by casperklein** [comment]
The following tasks have to be done (let me know if something is missing):

- Create a new branch for testing
- Remove target/fail2ban/jail.conf and go with the default debian config
- Implement things as discussed here
- Testing if fail2ban works as expected
- Review/Enhance bat tests

For the bat tests, I need support from someone else. Also testing should be ideally done by more people. The other stuff I should be able to take care of within the next weeks.

**Comment by polarathene** [comment]
Regarding the tests is that a concern with the bats test environment vs production environment differences? If so what kind of differences are of a concern?

**Comment by casperklein** [comment]
I am not sure I understand what you asking for exactly - sorry. My concern is pretty simple: I am not familiar with bats at the moment. So this part should be done by someone else with more knowledge.

**Comment by polarathene** [comment]
> I am not sure I understand what you asking for exactly - sorry.

Some things may not be as easy to test locally as they might be against a deployed instance. If you want to test fail2ban rules, that might also apply? (I'm not familiar with f2b much, just regarding IP or domain names)

> My concern is pretty simple: I am not familiar with bats at the moment. So this part should be done by someone else with more knowledge.

I had the same problem at the start of the year with my PRs. They're reasonably well documented I think with comments ([the bats tests](https://github.com/docker-mailserver/docker-mailserver/blob/master/test/security_tls_cipherlists.bats)), although mine differs a bit, I reworked [this one](https://github.com/docker-mailserver/docker-mailserver/blob/master/test/mail_ssl_manual.bats) too which is more like most tests.

I should be able to assist/guide you with any questions about putting a test together. As the docs are being reworked, this is also a contributing section that needs to be covered, so it's best to utilize your inexperience with writing the bats tests to get the questions and feedback to help knowing what needs to be covered :)

I'm not sure if tests can be run via Docker (since they start containers themselves), and I opted to run my tests in a temporary $5/month VPS. You need to be on Linux otherwise I think to run the tests, or use a VM (Ubuntu 20.04) that matches what [Github uses](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md).

I recall needing to run a command for cloning over the submodules (bats being one of them) to run, and another for initializing docker-mailserver for testing (dummy accounts) that `make` runs, as per the Github CI workflow we have does.

Other than that, it was just making a `test-name.bats` file like I linked to, that describes the docker-mailserver image configuration followed by some `@test` sections that run commands and assert the result is what you expect for success or failure conditions.

As you're testing new PR, you'd need to additionally build a new Docker image to test against (the tests use `${NAME}`) which defaults to `mailserver-testing:ci`.

I can help you with any of that, or any other questions you have :+1:

**Comment by casperklein** [comment]
@aendeavor What exact use case is not working for you?

I did some tests, however they all (both) succeed. I tried to send mails via tcp/25 & tcp/465. If I use wrong credentials (not existing user or existing user with wrong password) I get blocked.

**Comment by aendeavor** [comment]
> @aendeavor What exact use case is not working for you?
> 
> 
> 
> I did some tests, however they all (both) succeed. I tried to send mails via tcp/25 & tcp/465. If I use wrong credentials (not existing user or existing user with wrong password) I get blocked.
> 
> 
> 
> Edit: I remember I was having issues in the past too, but atm I am not able to reproduce them.

I've tried to setup F2B on the host with the Postfix aggressive jail. Absolutely no bans. I have not yet done the same for the container. Not sure which rules to set where, since I have not yet looked into it further.

I will post my setup later, I can't ATM. What's your setup?

**Comment by casperklein** [comment]
I did nothing on the host. In the container I've configured just `ENABLE_FAIL2BAN=1`.

**Comment by aendeavor** [comment]
How do you verify that?
In #1849, you said

> I see you use the DOCKER-USER chain. Using this chain, I was able to block the access. I didn't know that smile

What's up with this?

**Comment by casperklein** [comment]
I've tested adding a rule to the host docker-user chain manually, just to see if that works in principal. I've not tested a host-based fail2ban setup.

**Comment by casperklein** [comment]
What happens, if you connect to port 25/993/465 etc. and authenticate with wrong credentials multiple times. Then you should be banned. That works for me. Since you mentioned it does not work, I was interested what exactly did you try and what was the outcome.

**Comment by aendeavor** [comment]
> What happens, if you connect to port 25/993/465 etc. and authenticate with wrong credentials multiple times. Then you should be banned. That works for me. Since you mentioned it does not work, I was interested what exactly did you try and what was the outcome.

I will try tomorrow and report back. Thanks!

**Comment by aendeavor** [comment]
@casperklein Here is the part of `startup/setup-stack.sh` handling F2B:

https://github.com/docker-mailserver/docker-mailserver/blob/b46e48f278e66324d0c73f1c0b28d4bf52e14627/target/scripts/startup/setup-stack.sh#L1456-L1470

I double-checked and indeed, I have neither `fail2ban-fail2ban.cf` nor `fail2ban-jail.cf` in `/tmp/docker-mailserver` (unsurprisingly, since I did not setup these files in my `config/` directory). Is there a configuration step I missed? From what I know, the `jail.local` is a user-defined, obligatory file. Is this somewhere in the documentation? I'm rather confused right now...

**Comment by casperklein** [comment]
> Is there a configuration step I missed?

Nope, it's all fine.

At this point, I am not sure if we talked about the same thing ;-) You mentioned, F2B is not working for you. Your setup seems correct to me. Why do you conclude, your setup is not working? Don't you get banned, when wrong credentials are used repeatedly?

**Comment by aendeavor** [comment]
> At this point, I am not sure if we talked about the same thing ;-)

Haha, you're right. F2B is running, but I'm not sure it's doing what it is supposed to to :D

I already have a more or less rigorous Postfix setup (`smtpd_recipient_restrictions` and stuff...). I just never see that the Postfix jail has any bans:

``` BASH
$ ./setup.sh debug fail2ban
No IPs have been banned.
$ docker exec CONTAINER fail2ban-client status postfix
Status for the jail: postfix
|- Filter
|  |- Currently failed:	0
|  |- Total failed:	0
|  `- File list:	/var/log/mail/mail.log
`- Actions
   |- Currently banned:	0
   |- Total banned:	0
   `- Banned IP list:
```

**Comment by casperklein** [comment]
So when you configure your IMAP with a wrong user/password, you should be banned.

**Comment by aendeavor** [comment]
> So when you configure your IMAP with a wrong user/password, you should be banned.

You're right! I don't know why this did not come to my mind earlier... I just didn't test it. Thanks.

**Comment by casperklein** [comment]
Why mysterious? It's mentioned here: https://github.com/docker-mailserver/docker-mailserver/wiki/Configure-Fail2ban

**Comment by aendeavor** [comment]
> Why mysterious? It's mentioned here: https://github.com/docker-mailserver/docker-mailserver/wiki/Configure-Fail2ban

Yes, exactly. I couldn't find this either. No all is good :D

**Comment by BlockI0tChain** [comment]
If available, I'd gladly test any tag/configuration on my env.

**Comment by casperklein** [comment]
@aendeavor 

I am just wondering, if these are custom jails from you? I couldn't find any information, especially about the `postfix-block-spam` filter.

```
[postfix-block-spam]
enabled     = true
bantime     = 864000
filter      = postfix-block-spam
logpath     = /var/log/mail/mail.info

[dovecot-imap]
enabled     = true
port        = imap,imaps
filter      = dovecot-imap
logpath     = /var/log/mail/mail.info
```

fyi: you can use something like `10d` instead of `864000`

**Comment by aendeavor** [comment]
> @aendeavor
> 
> I am just wondering, if these are custom jails from you? 

Yes, they are custom :)

> fyi: you can use something like `10d` instead of `864000`

Nice to know - I will adjust my configuration

**Comment by casperklein** [comment]
I am going to close this issue soon, as almost all has been done:

- Now the Debian default configuration is used where possible
- An attacker is now blocked on all ports by default, instead of only the affected one. This can be reverted back via `config/fail2ban-jail.cf.`
- The packets are now *dropped* by default, however *reject* can also be configured via `$FAIL2BAN_BLOCKTYPE=drop|reject`.
- `FAIL2BAN_BLOCK_PERMANENT` was not implemented. This can be easily done, by setting a pretty long `bantime` in `config/fail2ban-jail.cf.`

The changes are now merged into the master branch, but have not been released yet.

@BlockI0tChain 

> If available, I'd gladly test any tag/configuration on my env.

That would be great. Please test and let us know, if you encounter any problems. I use this setup now for about a week and had no issues so far.

**Comment by BlockI0tChain** [comment]
### Some jails show filter errors, I guess could be misconfigured, I need to be sure of this, would someone please clear this up: 
What is the correct way to provide custom config for the jails?

**Comment by aendeavor** [comment]
> ### Some jails show filter errors, I guess could be misconfigured, I need to be sure of this, would someone please clear this up:
> 
> What is the correct way to provide custom config for the jails?

Could you please provide us with the error logs?

**Comment by BlockI0tChain** [comment]
@aendeavor testing the latest master #4127370 ofc  as agreed,  I'll post the logs after I'm sure I've configured everything as it should be. That's why I asked what is the correct way to provide configs for the jails in the latest? 
*   ./target/fail2ban/jail.local ?
*  ./config/fail2ban-jail.cf ?
*  ... ?

**Comment by wernerfred** [comment]
Does [this site from the docs](https://docker-mailserver.github.io/docker-mailserver/edge/config/security/fail2ban/) answer your question?

**Comment by BlockI0tChain** [comment]
My bad,  sry I missed it,  casperklein mentioned it as well above.

**Comment by chongma** [comment]
would be good to be able to whitelist a webmail ip like roundcube

**Comment by casperklein** [comment]
Currently this is defined:

```
# "ignoreip" can be a list of IP addresses, CIDR masks or DNS hosts. Fail2ban
# will not ban a host which matches an address in this list. Several addresses
# can be defined using space (and/or comma) separator.
ignoreip = 127.0.0.1/8
```

This can be overwritten/extended via `config/fail2ban-jail.cf`.

**Comment by BlockI0tChain** [comment]
#### My report as follows:
#### Used the same env as in [#1821](https://github.com/docker-mailserver/docker-mailserver/issues/1810#issuecomment-781521647) with the lastest master#4127370:



+ The Mailserver stack works fine as before,
+  F2B works partially:  
     +  Bots are getting detected and logged as before  ->
      + Now some of the bots are getting banned, most of them however aren't!!
        

Jail config :
+ [config/fail2ban-jail.cf](https://pastebin.com/a92jBLMC)

Can't put my finger on it yet. I suspect something is ill-set. 
All I can say for sure is that the whole cascade of the filter failures logged in the fail2ban status of the jails will most probably account for the ban enforcement issues. 
+ [fail2ban status](https://pastebin.com/Wt4nZkJR)
 
Please let me know if you can see something in the logs:


+ [docker log](https://pastebin.com/YpJKsU5L)
+ [fail2ban.log](https://pastebin.com/1qX1Judv)

**Comment by casperklein** [comment]
I've reopend #1821, lets continue there to keep this issue clean.
